### PR TITLE
Enable MOUSE, Unicode, extrakey, velocikey, backlight based on keymap keycodes

### PR DESCRIPTION
**Pull request by yanfali**
Scan the keycodes in a JSON file and enable rules.mk features based upon keycode usage. Trial feature: enable mouse, extrakeys, velocikey, backlight, and unicode

## Description
Today we force the makers to set what options should be enabled in their keyboard rules.mk based upon what they think their users will do with them. This creates a host of problems for configurator because we allow users to configure keycodes without taking into account whether the keyboard has that feature enabled or not.

### Pro
 * Keyboard no longer needs to specify all the features that a user could use at PR time
 * features the user does not want to use do not take up firmware space leaving more room for other features like macros in the future
 * we can now start adding support for things like macros based on the contents of the JSON file which is shared with configurator

### Cons
  * User could potentially enable too many features on a limited platform like AVR and create an unbuildable firmware combination

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/clueboard/california/keymaps/default/keymap.json [hunk 1]
```diff
@@ -1,7 +1,19 @@
...
-    "keyboard":"clueboard/california",
-    "layout":"LAYOUT",
-    "layers":[
-        ["KC_1","KC_2","KC_3","KC_4","KC_5","KC_6","KC_7","KC_8","KC_9","BL_STEP"]
+    "keyboard": "clueboard/california",
+    "layout": "LAYOUT",
+    "keymap": "fake_keymap",
+    "layers": [
+        [
+            "KC_MS_U",
```
Comment (yanfali) [inline keyboards/clueboard/california/keymaps/default/keymap.json]: this is just for testing purposes

**Comment by yanfali** [comment]
@skullydazed should we hide this behind a flag and use that exclusively on the API?

**Comment by Erovia** [comment]
I haven't had the time to check the code, but I really like the idea.
The "con" case should be handled gracefully with a nice error message and imho that's about it.

**Comment by skullydazed** [comment]
> @skullydazed should we hide this behind a flag and use that exclusively on the API?

My inclination is to have the API always behave the same as the default CLI options. This aids understanding and reproducibility. Is there a reason to break from that in this case?

**Comment by yanfali** [comment]
> > @skullydazed should we hide this behind a flag and use that exclusively on the API?
> 
> My inclination is to have the API always behave the same as the default CLI options. This aids understanding and reproducibility. Is there a reason to break from that in this case?

Not really. Just wondering if we make this experimental and test it out on the API first to not disrupt users. The thing I'm contemplating right now is defaulting all values to NO to save on flash space as we can detect their actual usage at run time.

**Comment by yanfali** [comment]
> I haven't had the time to check the code, but I really like the idea.
> The "con" case should be handled gracefully with a nice error message and imho that's about it.

Sorry, what's the "con" case?

**Comment by yanfali** [comment]
I've run this against the keymaps in the configurator repo, and 583 of the 620 compiled successfully into hex or bin files. I haven't investigated the failing ones, but my guess would be binaries being too large due to too many features. It's also possible that a certain number are just invalid.

**Comment by yanfali** [comment]
@skullydazed

**Comment by Erovia** [comment]
> > I haven't had the time to check the code, but I really like the idea.
> > The "con" case should be handled gracefully with a nice error message and imho that's about it.
> 
> Sorry, what's the "con" case?

> "User could potentially enable too many features on a limited platform like AVR and create an unbuildable firmware combination"

**Pull request closed without merging**